gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,58 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar, Union
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData, OutputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self) -> tf.Tensor:
+        """
+        Training loss for models that encapsulate their data.
+        """
+        return -self.maximum_a_posteriori_objective()
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
```

Comment:
tensorflow appear to use the word 'jit' as compiling for/with XLA. Isn't that different from `tf.function`. Should we be using a different word, one that they use, like `eager`? Really I'd prefer we leave it to the user to `tf.function` things and not have this argument at all

Reply:
`compile`?

Reply:
I've created a ticket re 'jit' #1373 . What do you think about my second point about users doing the `tf.function`-ing themselves. since we have the above closure, this just becomes
`tf.function(model.training_loss)` which is only 3 more characters than `model.training_loss_closure()`, and feel is mildly clearer

Reply:
Alas it's not so simple for the ExternalDataTrainingLossMixin (see the latest changes), so I think it needs to stay in there, and then for myself as a user I like having available the corresponding equivalent way of doing the same for models with internal data. (This also makes it easier to do e.g. `training_loss_closure(jit=number_of_calls>10)` or something like that, when you only want the compile overhead when you know you'll call multiple times...)

Reply:
can it not be added as a separate PR, especially since it's an optional argument and this PR's already 51 files?

Reply:
> can it not be added as a separate PR [...]?

What are you referring to? The `jit` flag? We definitely need it for the `training_loss_closure` for the ExternalData... version. It makes sense to me to have it for the InternalData... version, too, then. Most of the files are fixing use-cases, so separating it out into another PR doesn't seem very practical to me (lots of changes in this PR, plus lots of files changed in the other PR as well, for marginally removing the number of lines changed in this PR, at the cost of an overall much larger number of lines changed).

Reply:
why do you say we definitely need it? what functionality is it providing that we need in this PR?

Reply:
essentially I'm asking: what is the aim of this PR, and how does this functionality achieve that aim?

Reply:
my concern is that there's too much here to get the fundamentals in quick

Reply:
> do we have notebooks which can't be expressed easily without it?

Yes- they are all significantly shorter (and cleaner) now, don't you think?